First attempt to improve the description on the use cases. See issue 2 (#10)
x509-hbs/draft-gazdag-x509-hash-sigs: draft-gazdag-x509-hash-sigs.md

  RFC8410:
  RFC8411:
  RFC8708: #hsslms in cms
  MCGREW:
    title: State Management for Hash-Based Signatures
    author:
    -
      ins: D. McGrew
    -
      ins: P. Kampanakis
    -
      ins: S. Fluhrer
    -
      ins: S. Gazdag
    -
      ins: D. Butin
    -
      ins: J. Buchmann
    date: 2016-11-02
  NISTSP:

>     title: Recommendation for Stateful Hash-Based Signature Schemes

Reviewer: https://doi.org/10.6028/NIST.SP.800-208

    author:
    -
      ins: National Institute of Standards and Technology (NIST)
    date: 2020-10-29
  NSA:

>     title: Commercial National Security Algorithm Suite 2.0 (CNSA 2.0) Cybersecurity Advisory (CSA)

Reviewer: https://media.defense.gov/2022/Sep/07/2003071834/-1/-1/0/CSA_CNSA_2.0_ALGORITHMS_.PDF

Hash-Based Signature (HBS) Schemes combine Merkle trees with One/Few Time
Signatures (OTS/FTS) in order to provide digital signature schemes that remain
secure even when quantum computers become available. Their security is well
understood and depends only on the security of the underlying hash function. As
such they can serve as an important building block for quantum computer
resistant information and communication technology.

> The private key of HSS, XMSS and XMSS^MT is a finite collection of OTS keys, hence only a limited
> number of messages can be signed and the private key's state must be updated and persisted
> after signing to prevent reuse of OTS keys. Due to this statefulness of the private key
> and the limited number of signatures that can be created, these signature algorithms might
> not be appropriate for use in interactive protocols. While the right selection of algorithm
> parameters would allow a private key to sign a virtually unbounded number of messages (e.g. 2^60),
> this is at the cost of a larger signature size and longer signing time. Since these algorithms
> are already known to be secure against quantum attacks, and because roots of trust are generally
> long-lived and can take longer to be deployed than end-entity certificates, these signature
> algorithms are more appropriate to be used in some use case scenarios.

Reviewer 2: ```suggestion
The private key of HSS, XMSS and XMSS^MT is a finite collection of OTS keys, hence only a limited
number of messages can be signed and the private key's state must be updated and persisted after signing to prevent reuse of OTS keys. While the right selection of algorithm parameters would allow a private key to sign a virtually unbounded number of messages (e.g. 2^60), this is at the cost of a larger signature size and longer signing time. Due to the statefulness of the private key of HSS, XMSS and XMSS^MT and the limited number of signatures that can be created, these signature algorithms might not be appropriate for use in interactive protocols. However, in some use case scenarios the deployment of these signature algorithms may be appropriate. Such use cases are described and discussed later in {{use-cases-hbs-x509}}.
```

trusted code signing certificates are useful.
The private key of HSS, XMSS and XMSS^MT is a finite collection of OTS keys, hence only a limited
number of messages can be signed and the private key's state must be updated and persisted
after signing to prevent reuse of OTS keys. Due to this statefulness of the private key
and the limited number of signatures that can be created, these signature algorithms might
not be appropriate for use in interactive protocols. While the right selection of algorithm
parameters would allow a private key to sign a virtually unbounded number of messages (e.g. 2^60),
this is at the cost of a larger signature size and longer signing time. Since these algorithms
are already known to be secure against quantum attacks, and because roots of trust are generally
long-lived and can take longer to be deployed than end-entity certificates, these signature
algorithms are more appropriate to be used in some use case scenarios.

The private key of SPHINCS+ is a finite but very large collection of FTS keys
and hence stateless. This typically comes at the cost of larger signatures
compared to the stateful HBS variants. Thus SPHINCS+ is suitable for more

> use-cases if the signature sizes fit the requirements.

Reviewer 2: Einheitlich ohne Bindestrich.

use cases if the signature sizes fit the requirements.

# Conventions and Definitions

{::boilerplate bcp14-tagged}

The parameter 'n' is the security parameter, given in bytes. In practice this
is typically aligned to the standard output length of the hash function in use,

> i.e. either 24, 32 or 64 bytes. The height of a single tree is typically given

Reviewer 2: Add 16 Bytes as possible value for "n" (see SPHINCS+-128)

> # Use Cases of HBS in X.509

Reviewer 2: ```suggestion
# Use Cases of HBS in X.509 {#use-cases-hbs-x509}
```
Add section label for dynamic referencing.

> a secure instantiation of HBS may always be built as long as a cryptographically

Reviewer 2: ```suggestion
and a secure instantiation of HBS may always be built as long as a cryptographically
```
@exception42 : I've taken Kaveh's comment to submit a suggestion. Please commit if ok.

> can be built. The major negative aspect is the statefullness of several HBS.

Reviewer 2: ```suggestion
can be built. The major negative aspect is the statefulness of several HBS.
```
@exception42 : Commit if ok.

> security incidents like signature forgery [MCGREW], [NISTSP].
> 
> As a secure environment MUST be used for key generation and key management,
> HBS are still a proper solution for root CA certificates. HBS MAY be used for
> subordinate CA certificates, but special consideration MUST be taken into account
> for proper key management. HBS MUST NOT be used for end-entity certificates.

Reviewer 2: ```suggestion
security incidents like signature forgery [MCGREW], [NISTSP]. Therefore, for stateful HBS, a secure environment MUST be used for key generation and key management.

Note that, in general, root CAs offer such a secure environment and the number of issued signatures (including signed certificates and CRLs) is often moderate due to the fact that many root CAs delegate OCSP services or the signing of end-entity certificates to other entities (such as subordinate CAs) that use stateless signature schemes. Therefore, many root CAs should be able to handle the required state management, and stateful HBS offer a viable solution.

HBS MAY also be used by subordinate CAs for issuing certificates, but special and careful consideration MUST be taken into account for proper key management. HBS MUST NOT be used for end-entity certificates.
```

> such they can serve as an important building block for quantum computer

Reviewer: Im vorhergehenden Satz muss "There" ersetzt werden durch "Their"

Reviewer 2: Typo behoben

resistant information and communication technology.

The private key of HSS, XMSS and XMSS^MT is a finite collection of OTS keys, hence only a limited
number of messages can be signed and the private key's state must be updated and persisted

> after signing to prevent reuse of OTS keys. Due to thise statefulness of the private key

Reviewer 2: ```suggestion
after signing to prevent reuse of OTS keys. Due to this statefulness of the private key
```

and the limited number of signatures that can be created, these signature algorithms might
not be appropriate for use in interactive protocols. While the right selection of algorithm
parameters would allow a private key to sign a virtually unbounded number of messages (e.g. 2^60),
this is at the cost of a larger signature size and longer signing time. Since these algorithms
are already known to be secure against quantum attacks, and because roots of trust are generally
long-lived and can take longer to be deployed than end-entity certificates, these signature

> algorithms are more appropriate to be used in some use case scenarios.

Reviewer: Ich finde den Aufbau dieses Absatzes nicht stimmig:
Der 1. und 3. Satz beschreibt HBS und listet einige
Eigenschaften auf.
Der 2. und 4. Satz sagt etwas über mögliche use-cases.
Außerdem: Dass HBS quantensicher ist, steht schon im
vorherigen Abschnitt. Ich würde es hier nicht mehr
wiederholen, da es im "use cases-Kapitel auch erwähnt wird.
https://github.com/x509-hbs/draft-gazdag-x509-hash-sigs/pull/10/commits/95d5bd7914247daeac0bf3c6e03acd541ae8223f?di…
2/615.06.23, 16:26
First attempt to improve the description on the use cases. See issue 2 by exception42 · Pull Request #10 · x5…
Mein Gegenvorschlag:
The private key of HSS, XMSS and XMSS^MT is a finite collection
of OTS keys, hence only a limited
number of messages can be signed and the private key's state
must be updated and persisted
after signing to prevent reuse of OTS keys. While the right selection
of algorithm
parameters would allow a private key to sign a virtually unbounded
number of messages (e.g. 2^60),
this is at the cost of a larger signature size and longer signing time.
Due to the statefulness of the private key of HSS, XMSS and
XMSS^MT and the limited number of signatures that can be
created, these signature algorithms might not be appropriate for
use in interactive protocols. However, in some use case scenarios
the deployment of these signature algorithms may be appropriate.
Such use cases are described and discussed in the next section.

The private key of SPHINCS+ is a finite but very large collection of FTS keys

> and hence stateless. This typically comes at the cost of larger signatures

Reviewer: Im Satz "Thus SPHINCS+..." wird "use-cases" geschrieben. Wir sollten uns entscheiden, ob wir es mit oder ohne Bindestrich schreiben und konsistent bleiben :-)

> by the parameter 'h'. The number of levels of trees is either called 'L' (HSS)

Reviewer: 16 bytes ist auch möglich (z.B. SPHINCS+-128s)

or 'd' (XMSS, XMSS^MT, SPHINCS+).

# Use Cases of HBS in X.509

As many cryptographic algorithms that are considered to be quantum-resistant,
HBS have several pros and cons regarding their practical usage. On the positive side
they are considered to be secure against a classical as well as a quantum adversary,

> a secure instantiation of HBS may always be built as long as a cryptographically

Reviewer: ", a secure instantiation of HBS" ersetzen durch ", and a secure instantiation of HBS"

secure hash function exists. While key generation and signature generation may

> take longer than classical alternatives, fast and minimal verification routines

Reviewer 2: Maybe add something like:
- In comparison to other post-quantum signature schemes stateful ones can have (small signature + public key) sizes.

Reviewer: Nach dem Satz "... secure hash function exists." würde ich Folgendes einfügen:

Moreover, HBS offer small public key sizes, and, in comparison to other post-quantum signature schemes, the stateful HBS offer relatively small signauture (for certain parameter sets).

> can be built. The major negative aspect is the statefullness of several HBS.

Reviewer: Nur ein "l" bei "statefullness"? Bin mir da nicht so sicher. Man schreibt glaub ich auf jeden Fall "stateful".

Private keys always have to be handled in a secure manner, but stateful HBS
necessitate a special treatment of the private key in order to avoid
security incidents like signature forgery [MCGREW], [NISTSP].

As a secure environment MUST be used for key generation and key management,

> HBS are still a proper solution for root CA certificates. HBS MAY be used for

Reviewer: Mein Vorschlag: 

... in order to avoid security incidents like signature forgery [MCGREW], [NISTSP]. Therefore, for stateful HBS, a secure environment MUST be used for key generation and key management.

Note that, in general, root CAs offer such a secure environment and the number of issued signatures (including signed certificates and CRLs) is often moderate due to the fact that many root CAs delegate OCSP services or the signing of end-entity certificates to other entities (such as subordinate CAs) that use non-stateful signature schemes for their signatures. Therefore, many root CAs should be able to handle the required state management, and stateful HBS are still a proper solution for usage there. 

HBS MAY also be used for subordinate CA certificates, but special and careful consideration MUST be taken into account for proper key management. HBS MUST NOT be used for end-entity certificates.

We include some further remarks on this issues in the section on security considerations.
